### DISCUSSION TITLE
Transitive CPATH

### DISCUSSION THREAD

**Discussion by HadrienG2** [spack/spack#37287] [Q&A]
So, [this interesting PR just arrived for the ROOT package](https://github.com/spack/spack/pull/37278), and it raises an interesting question.

Assume you have a C/++ program A which includes a header from dependency B, and inside of that header, there is an include from transitive dependency C. But project A does not directly use dependency C, it's just used somewhere in the implementation of B.

How does one get this to work in Spack's single-package-per-prefix directory layout, which means somehow the headers of C must be in CPATH at the time where A will be built?

- Should A add C as a dependency? That would not feel right, since A does not directly use C, and should not need to care that B is implemented using C.
- Should the package.py B add the CPATH of C to the CPATH that it exposes to dependents, via `setup_dependent_build_environment` and friends? This is what #37278 is going for, but if applied systematically it would mean a large amount of extra packaging work for maintainers of package B, who would now need to track all public headers to see which transitive deps appear in public headers and make sure their CPATHs are propagated appropriately?
- Should spack just automatically add the CPATH of all dependencies to the CPATH of dependent builds? That sounds like it could be a good "just works" default policy, but perhaps people would like to inhibit it sometimes due to header path collisions?

Most importantly, however, is this really the first time that this problem is explored with spack? Or is it well known and does it have a well accepted solution?

**Comment by haampie** [comment]
I think in order of preference:

1. Rely on the build system (Spack sets `CMAKE_PREFIX_PATH` / `PKG_CONFIG_PATH` for transitive deps)
2. Rely on the compiler wrapper (it injects -I flags for transitive deps -- see `SPACK_TEST_COMMAND=dump-args spack env build-env <your spec> -- cc foo.c`). Include paths for transitive link deps are added automatically (through `spec["dep_of_dep"].headers.directories`)
3. Modify package.py

**Reply by jmcarcell** [reply]
In our case, B and C are built with spack and provided in a closed environment by exporting all the paths. Then a user comes, gets all the paths and without using spack compiles A so 1 and 2 can't be done I imagine, can we do better than 3 with our workflow?

**Reply by haampie** [reply]
I think you have two options there:

1. Use an environment view, so all headers end up in a common dir: `<view>/include/*`
2. Set custom prefix inspections in `modules.yaml`; this affects `spack load` / `module load`. The format is: `modules:prefix_inspections:relative_path_in_package_prefix:[VARIABLE_1, VARIABLE_2]`

So, for example `spack config add modules:prefix_inspections:include:[CPATH]` should set `CPATH=<prefix>/include` on `spack load` for each spec that has an include dir.

1 and 2 can also be combined: `spack env activate` would then set `CPATH` to `<view>/include`
